dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/deduplication/DeduplicationControllerTest.java

```diff
@@ -0,0 +1,169 @@
+    private ObjectMapper objectMapper;
+
+    private TrackedEntityInstance origin;
+
+    private TrackedEntityInstance duplicate;
+
+    @BeforeEach
+    public void setUp()
+    {
+        OrganisationUnit orgUnit = createOrganisationUnit( CodeGenerator.generateUid() );
+        dbmsManager.save( orgUnit );
+
+        origin = createTrackedEntityInstance( orgUnit );
+        duplicate = createTrackedEntityInstance( orgUnit );
+
+        dbmsManager.save( origin );
+        dbmsManager.save( duplicate );
+    }
+
+    @Test
+    void shouldPostPotentialDuplicateWhenTrackedEntitiesExist()
+        throws Exception
+    {
+        PotentialDuplicate potentialDuplicate = new PotentialDuplicate( origin.getUid(), duplicate.getUid() );
+        assertStatus( HttpStatus.OK, POST( ENDPOINT, objectMapper.writeValueAsString( potentialDuplicate ) ) );
+    }
+
+    @Test
+    void shouldThrowPostPotentialDuplicateWhenMissingDuplicateTeiInPayload()
+        throws Exception
+    {
+        PotentialDuplicate potentialDuplicate = new PotentialDuplicate( origin.getUid(), null );
+        assertStatus( HttpStatus.BAD_REQUEST, POST( ENDPOINT, objectMapper.writeValueAsString( potentialDuplicate ) ) );
+    }
+
+    @Test
+    void shouldThrowPostPotentialDuplicateWhenMissingOriginTeiInPayload()
+        throws Exception
+    {
+        PotentialDuplicate potentialDuplicate = new PotentialDuplicate( null, duplicate.getUid() );
+        assertStatus( HttpStatus.BAD_REQUEST, POST( ENDPOINT, objectMapper.writeValueAsString( potentialDuplicate ) ) );
+    }
+
+    @Test
+    void shouldThrowBadRequestWhenPutPotentialDuplicateAlreadyMerged()
+    {
+        PotentialDuplicate potentialDuplicate = new PotentialDuplicate( origin.getUid(), duplicate.getUid() );
+        potentialDuplicate.setStatus( DeduplicationStatus.MERGED );
+        save( potentialDuplicate );
+
+        assertStatus( HttpStatus.BAD_REQUEST,
+            PUT( ENDPOINT + potentialDuplicate.getUid() + "?status=" + DeduplicationStatus.INVALID.name() ) );
+    }
+
+    @Test
+    void shouldThrowBadRequestWhenPutPotentialDuplicateToMergedStatus()
+    {
+        PotentialDuplicate potentialDuplicate = potentialDuplicate( origin.getUid(), duplicate.getUid() );
+        assertStatus( HttpStatus.BAD_REQUEST,
+            PUT( ENDPOINT + potentialDuplicate.getUid() + "?status=" + DeduplicationStatus.MERGED.name() ) );
+    }
+
+    @Test
+    void shouldUpdatePotentialDuplicateWhenPotentialDuplicateExistsAndCorrectStatus()
+    {
+        PotentialDuplicate potentialDuplicate = potentialDuplicate( origin.getUid(), duplicate.getUid() );
+        assertStatus( HttpStatus.OK,
+            PUT( ENDPOINT + potentialDuplicate.getUid() + "?status=" + DeduplicationStatus.INVALID.name() ) );
+    }
+
+    @Test
+    void shouldThrowBindExceptionWhenGetPotentialDuplicateWithInvalidStatus()
+    {
+        assertStatus( HttpStatus.BAD_REQUEST,
+            GET( ENDPOINT + "?status=" + "invalidStatus" ) );
```

Comment:
@enricocolasante should we test these BindExceptions and MethodArgumentTypeMismatchExceptions in individual controllers? Or is it simply duplicating tests you already have elsewhere?